grep/hemartin/grep.go

```diff
@@ -0,0 +1,94 @@
+package grep
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+	"regexp"
+)
+
+const printLinesFlag = "-n"
+const printNamesFlag = "-l"
+const caseInsensitiveFlag = "-i"
+const invertFlag = "-v"
+const matchLinesFlag = "-x"
+
+func Search(pattern string, flags, files []string) []string {
+	result := []string{}
+
+	printLines := sliceContains(flags, printLinesFlag)
+	caseInsensitive := sliceContains(flags, caseInsensitiveFlag)
+	printNames := sliceContains(flags, printNamesFlag)
+	matchLines := sliceContains(flags, matchLinesFlag)
+	invert := sliceContains(flags, invertFlag)
+
+	regexFlags := ""
+	if caseInsensitive {
+		regexFlags += "(?i)"
+	}
+
+	if matchLines {
+		pattern = fmt.Sprintf("^%s$", pattern)
+	}
+
+	regexPattern := fmt.Sprintf(`%s%s`, regexFlags, pattern)
+
+	compiledPattern := regexp.MustCompile(regexPattern)
+	filesContents := readFiles(files)
+	for _, fileName := range files {
+		lines := filesContents[fileName]
+		for lineIndex, line := range lines {
+			if string(line) == "" {
+				continue
+			}
+
+			match := compiledPattern.Match(line)
+			if (match && !invert) || (!match && invert) {
```

Comment:
Another way to do the same `if invert { match = !match }; if match {...}`. If you prefer more unreadable then `if match != invert {...}` (this is basically a XOR).